### ISSUE TITLE
A novel syntax-aware automatic graphics code generation with attention-based deep neural network

### ISSUE THREAD

**Issue by e4exp** [e4exp/paper_manager_abstract#152]
* https://www.sciencedirect.com/science/article/pii/S1084804520301107?casa_token=m-6bYDcMQOkAAAAA:ToF1y58cH1keUI-JptsF9PKs1vox2WDe_Tut4rAZrzXJpfCsb5xXr3tvt3pVbpT4Q56Zt-L4nUs
* Journal of Network and Computer Applications Volume 161, 1 July 2020, 102636

最近のディープラーニングの進歩により、エンコーダー・デコーダーフレームワークによってグラフィカル・ユーザー・インターフェース（GUI）を自動的にコードに変換することが可能になりました。
このフレームワークは一般的に、深層畳み込みニューラルネットワーク(CNN)を用いて画像の特徴を抽出し、それをリカレントニューラルネットワーク(RNN)をベースにしたコードジェネレータによって数百のコードトークンに変換する。
しかし、このフレームワークの実装には、GUIやドメイン指定言語（DSL）コードに含まれる情報をいかに活用するかという課題と、生成されたDSLコードをいかに構文規則に適合させるかという課題があります。
GUIやDSLコードに含まれる情報をフルに活用するために、まず、視覚的な注意喚起機能（CNNで抽出）とDSLの注意喚起機能（LSTMで抽出）を統合したHGui2Codeというモデルを提案する。
さらに，ON-LSTMネットワークを利用して，構文的に正しいDSLコードを生成する新しいモデルであるSGui2Codeを提案する．
HGui2Codeは意味情報に注目しているのに対し、SGui2Codeは文法規則に注目している。広範な実験結果から、我々のモデルはウェブデータセット上で最先端の手法を凌駕しており、HGui2Codeモデルでは5.5％、SGui2Codeモデルでは1.5％の精度向上を達成した。
IOSやAndroidのデータセットでは大きな効果は得られていないが、本モデルで生成されたDSLコードは、対応するGUIのコンポーネントのレイアウトに非常に近いものであることが示された。

**Comment by e4exp** [comment]
# A novel syntax-aware automatic graphics code generation with attention-based deep neural network

* paper 
    * https://www.sciencedirect.com/science/article/pii/S1084804520301107?casa_token=m-6bYDcMQOkAAAAA:ToF1y58cH1keUI-JptsF9PKs1vox2WDe_Tut4rAZrzXJpfCsb5xXr3tvt3pVbpT4Q56Zt-L4nUs
* github
    * 
* project
    * 

![スクリーンショット 2020-09-08 15 12 29](https://user-images.githubusercontent.com/44970465/92439403-beda3e00-f1e5-11ea-9a41-aa9842e6731e.png)


## まとめ

* **どんなもの？** 
    * GUI画像からDSLを生成するタスクで，先行研究(pix2code)の改善モデルを2種類提案．１つは画像encoderとDSL encoderにそれぞれattentionを追加したHGui2Code, もう一つはsyntax情報を考慮して文法に則った生成が可能なON-LSTMをDSL encoderとdecoderに使用したSGui2Code．
    * webドメインで，HGui2Codeが5.5%，SGui2Codeが1.5%高いaccuracyを達成
* **先行研究と比べてどこがすごい？**
    * webドメインでのerror低減．(iOS,Androidデータセットでは改善なし)
* **技術や手法のキモはどこ？**
    * HGui2Code
        * 画像に対してのvisual attentionと，DSLに対してのattentionを独立に計算し，concatで結合してdecoder LSTMへの入力にしている
    * SGui2Code
        * ON-LSTMをDSLのencoder LSTMに使用し，image featureと合わせた特長を受け取るdecoderもON-LSTMにしている．こちらはattentionは使用しない．
* **どうやって有効だと検証した？**
    * ABHD, pix2codeとの性能比較．visual attentionのみを使用したAGui2Codeというベースラインモデルも比較している
    * データセットはpix2codeが公開しているDSL(web, android, iOS)
* **議論はある？**    
    * webだけが改善して，他の２つのデータでは改善しない理由として，他２つのデータは似たものが多いため，DSLと画像の関係を捉えられていないとしている
    * visual attentionだけを使用したAGui2Codeのほうが，DSLにもattentionするHGui2Codeよりも結果がよい．これはDSLのattentionとVisual attentionが独立に計算されるため，それらのattention間のalingmentができていないためとしている(attention on attentionが使えるのでは，と思った)
    * ON-LSTMはneuronに保存される情報のライフサイクルをコントロールできるので，通常のLSTMより長期の依存関係に強いとしている
    * (attentionとON-LSTMを併用しないのはなぜ？)
* **次に読むべき論文は？**
    * Neelakantan, A., Le, Q.V., Sutskever, I., 2015. Neural Programmer: Inducing Latent Programs with Gradient descent[J] arXiv preprint. 1511.04834. 
        * プログラム生成の先行研究．組み合わせプログラム
    * Balog, M., Gaunt, A.L., Brockschmidt, M., et al., 2016. Deepcoder: Learning to Write programs[J] arXiv preprint. 1611.01989. 
        * プログラム生成の先行研究．input-outputcompetition-style programmingの解法
    * Shen, Y., Tan, S., Sordoni, A., et al., 2018. Ordered Neurons: Integrating Tree Structures into Recurrent Neural Networks[J] arXiv preprint. 1810.09536.
        * 文法規則に則ったcode生成が可能なLSTM．SGui2Codeが使用している
    * Kant, N., 2018. Recent Advances in Neural Program Synthesis[J] arXiv preprint. 1802.02353. 
        * プログラム生成手法のレビュー論文．
    * Deng, Y., Kanervisto, A., Rush, A.M., 2016. What You Get Is what You See: A Visual Markup decompiler[J], vol. 10, pp. 32–37 arXiv preprint. 1609.04938. 
        * 画像をデコンパイルしてrepresentation markerにする．gui code generationの先行研究
    * Ellis, K., Ritchie, D., Solar-Lezama, A., et al., 2018. Learning to infer graphics programs from hand-drawn images[C]. In: Advances in Neural Information Processing Systems, pp. 6060–6069. 
        * 手書き文字をtexサブセットのgraphical programモデルに変換する．gui code generationの先行研究
    * Zhu, Z., Xue, Z., Yuan, Z., 2018. Automatic Graphics Program Generation Using Attention-Based Hierarchical Decoder[J] arXiv preprint. 1810.11536. 
        * 画像の特徴を，block levelとtoken levelの２レベルでdecodeする．gui code generationの先行研究
    * Liu, X., Kong, X., Liu, L., et al., 2018. TreeGAN: syntax-aware sequence generation with generative adversarial networks[C]. In: 2018 IEEE International Conference on Data Mining (ICDM). IEEE, pp. 1140–1145. 
        * sequence 生成にcontext free grammerを組み込んだGAN．SQLクエリを自動生成．
    * Shen, Y., Tan, S., Sordoni, A., et al., 2018. Ordered Neurons: Integrating Tree Structures into Recurrent Neural Networks[J] arXiv preprint. 1810.09536. 
        * ON-LSTM．階層構造をモデリングするためにneuronに順序をもたせた


* **その他**
    *
